src/main/java/org/embulk/input/gcs/SingleFileProvider.java

```diff
@@ -52,10 +52,14 @@ public InputStream openNext() throws IOException
         String key = iterator.next();
         Storage.Objects.Get getObject = client.objects().get(bucket, key);
         File tempFile = Exec.getTempFileSpace().createTempFile();
+        GcsInputStreamReopener reopener = new GcsInputStreamReopener(tempFile, client, bucket, key, maxConnectionRetry);
         try (BufferedOutputStream outputStream = new BufferedOutputStream(new FileOutputStream(tempFile))) {
             IOUtils.copy(getObject.executeMediaAsInputStream(), outputStream);
         }
-        return new ResumableInputStream(new BufferedInputStream(new FileInputStream(tempFile)), new GcsInputStreamReopener(tempFile, client, bucket, key, maxConnectionRetry));
+        catch (Exception ex) {
+            reopener.reopen(0, ex);
```

Comment:
@sakama i'm still confusing about this part. When `IOUtils.copy` method might throw `IOException`, can the plugin read the temp file by reopener correctly? The temp file might be broken (or the plugin could read it partially though).

Reply:
Won't works?
In this part, plugin will write(not read) at temp file.
And plugin will do following processing in the RetryExcecutor.
* Try to get [InputStream](https://github.com/embulk/embulk-input-gcs/blob/fix-retry-logic-for-ioutils-copy/src/main/java/org/embulk/input/gcs/SingleFileProvider.java#L103) again with `getObject.executeMediaAsInputStream()`
* Try to open new [FileOutputStream](https://github.com/embulk/embulk-input-gcs/blob/fix-retry-logic-for-ioutils-copy/src/main/java/org/embulk/input/gcs/SingleFileProvider.java#L105) and override temp file.

At last, plugin will read from temp file and return it.
https://github.com/embulk/embulk-input-gcs/blob/fix-retry-logic-for-ioutils-copy/src/main/java/org/embulk/input/gcs/SingleFileProvider.java#L62

Reply:
> If we could use offset parameter in the reopen method to download GCS file, it's better to implement it in the reopen method.

Unfortunately, `getObject.executeMediaAsInputStream()` and other methods in GCS SDK doesn't allow range request to download. 
AWS S3 is possible, but.
https://github.com/embulk/embulk-input-s3/blob/v0.2.14/embulk-input-s3/src/main/java/org/embulk/input/s3/AbstractS3FileInputPlugin.java#L301

So plugin need to download all ranges of file data every time from GCS bucket.

Reply:
Yes, that's what I was guessing. 

> what do you think that we'd better to change that IOUtils.copy is called in openNext method only? 

What do you think about my above question in the previous comment? I think that the plugin doesn't need to download GCS files in `reopen` method and it could download (or retry downloading) them in `IOUtils.copy` method only.

Reply:
I changed implementation to use `RetryExecutor` directory instead of `GcsInputStreamReopener` 54b0f37

Reply:
Exception happens while `IOUtils.copy()` will be also retried 2f3bf2d